tortoise/verifying_tortoise.go

```diff
@@ -809,27 +809,164 @@ func (t *turtle) handleLayerBlocks(ctx *tcontext, layerID types.LayerID) error {
 		return fmt.Errorf("unable to read contents of layer %v: %w", layerID, err)
 	}
 	if len(layerBlocks) == 0 {
-		// nothing to do
 		t.logger.WithContext(ctx).Warning("cannot process empty layer block list")
 		return nil
 	}
 	return t.processBlocks(ctx, layerBlocks)
 }
 
+func (t *turtle) verifyingTortoise(ctx *tcontext, logger log.Log, lid types.LayerID) (map[types.BlockID]bool, error) {
+	localOpinion, err := t.getLocalOpinion(ctx, lid)
+	if err != nil {
+		return nil, err
+	}
+	if len(localOpinion) == 0 {
+		// warn about this to be safe, but we do allow empty layers and must be able to verify them
+		logger.With().Warning("empty vote vector for layer", lid)
+	}
+
+	contextualValidity := make(map[types.BlockID]bool, len(localOpinion))
+
+	// Count the votes of good blocks. localOpinionOnBlock is our local opinion on this block.
+	// Declare the vote vector "verified" up to position k if the total weight exceeds the confidence threshold in
+	// all positions up to k: in other words, we can verify a layer k if the total weight of the global opinion
+	// exceeds the confidence threshold, and agrees with local opinion.
+	for bid, localOpinionOnBlock := range localOpinion {
+		// count the votes of the input vote vector by summing the voting weight of good blocks
+		logger.With().Debug("summing votes for candidate layer block",
+			bid,
+			log.FieldNamed("layer_start", lid.Add(1)),
+			log.FieldNamed("layer_end", t.Last),
+		)
+		sum, err := t.sumVotesForBlock(ctx, bid, lid.Add(1), func(votingBlockID types.BlockID) bool {
```

Comment:
I think extracting the anonymous function would make the function a bit more readable